lib/std/fs/Dir.zig

```diff
@@ -1265,21 +1265,14 @@ pub fn realpathW(self: Dir, pathname: []const u16, out_buffer: []u8) ![]u8 {
     };
     defer w.CloseHandle(h_file);
 
-    // Use of MAX_PATH_BYTES here is valid as the realpath function does not
-    // have a variant that takes an arbitrary-size buffer.
-    // TODO(#4812): Consider reimplementing realpath or using the POSIX.1-2008
-    // NULL out parameter (GNU's canonicalize_file_name) to handle overelong
-    // paths. musl supports passing NULL but restricts the output to PATH_MAX
-    // anyway.
-    var buffer: [fs.MAX_PATH_BYTES]u8 = undefined;
-    const out_path = try posix.getFdPath(h_file, &buffer);
-
-    if (out_path.len > out_buffer.len) {
+    var wide_buf: [w.PATH_MAX_WIDE]u16 = undefined;
+    const wide_slice = try w.GetFinalPathNameByHandle(h_file, .{}, &wide_buf);
+    var big_out_buf: [fs.MAX_PATH_BYTES]u8 = undefined;
+    const end_index = try std.unicode.utf16leToUtf8(&big_out_buf, wide_slice);
```

Comment:
Windows file paths may be invalid UTF-16. They can be arbitrary bytes technically

Reply:
Yes, and this bug is present in many places throughout the standard library. @squeek502 is working on a comprehensive fix. The cross-platform error sets will likely not change since WASI made the decision that file paths are required to be well-formed UTF-8, which means it is possible to create file paths on every operating system that are inaccessible by WASI.

The code you are commenting on is inlined logic; it is not introducing any new assumptions.

Reply:
Actually the previous implementation did `catch unreachable` if the file path contained invalid utf-8, so this is at least a step in the right direction. Watch for that PR from squeek for the proper fix.